File: scripts/reference-solutions/subarray-sum-in-range.py
import sys
from bisect import bisect_left, bisect_right


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    low = int(data[1])
    high = int(data[2])
    nums = [int(x) for x in data[3 : 3 + n]]

    # Prefix sums: a subarray (i, j] has sum prefix[j] - prefix[i].
    prefix = [0] * (n + 1)
    for i in range(n):
        prefix[i + 1] = prefix[i] + nums[i]

    # Coordinate-compress the prefix values and count, for each j, how many
    # earlier prefix[i] fall in [prefix[j] - high, prefix[j] - low] using a
    # Fenwick tree. Overall O(n log n).
    sorted_prefix = sorted(set(prefix))
    size = len(sorted_prefix)
    tree = [0] * (size + 1)

    def update(pos0):
        i = pos0 + 1
        while i <= size:
            tree[i] += 1
            i += i & (-i)

    def prefix_count(count):
        i = count
        total = 0
        while i > 0:
            total += tree[i]
            i -= i & (-i)
        return total

    ans = 0
    update(bisect_left(sorted_prefix, prefix[0]))
    for j in range(1, n + 1):
        lo = prefix[j] - high
        hi = prefix[j] - low
        right_rank = bisect_right(sorted_prefix, hi)
        left_rank = bisect_left(sorted_prefix, lo)
        ans += prefix_count(right_rank) - prefix_count(left_rank)
        update(bisect_left(sorted_prefix, prefix[j]))

    print(ans)


main()

File: scripts/reference-solutions/subarray-sum-in-range.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  long long n, low, high;
  cin >> n >> low >> high;
  vector<long long> nums(n);
  for (long long& value : nums) {
    cin >> value;
  }

  vector<long long> prefix(n + 1, 0);
  for (long long i = 0; i < n; ++i) {
    prefix[i + 1] = prefix[i] + nums[i];
  }

  vector<long long> sortedPrefix(prefix.begin(), prefix.end());
  sort(sortedPrefix.begin(), sortedPrefix.end());
  sortedPrefix.erase(unique(sortedPrefix.begin(), sortedPrefix.end()), sortedPrefix.end());
  int size = static_cast<int>(sortedPrefix.size());
  vector<long long> tree(size + 1, 0);

  auto update = [&](int pos0) {
    for (int i = pos0 + 1; i <= size; i += i & (-i)) {
      tree[i] += 1;
    }
  };
  auto prefixCount = [&](int count) -> long long {
    long long total = 0;
    for (int i = count; i > 0; i -= i & (-i)) {
      total += tree[i];
    }
    return total;
  };
  auto positionOf = [&](long long value) {
    return static_cast<int>(lower_bound(sortedPrefix.begin(), sortedPrefix.end(), value) -
                            sortedPrefix.begin());
  };

  long long ans = 0;
  update(positionOf(prefix[0]));
  for (long long j = 1; j <= n; ++j) {
    long long lo = prefix[j] - high;
    long long hi = prefix[j] - low;
    int rightRank =
        static_cast<int>(upper_bound(sortedPrefix.begin(), sortedPrefix.end(), hi) -
                         sortedPrefix.begin());
    int leftRank = static_cast<int>(lower_bound(sortedPrefix.begin(), sortedPrefix.end(), lo) -
                                    sortedPrefix.begin());
    ans += prefixCount(rightRank) - prefixCount(leftRank);
    update(positionOf(prefix[j]));
  }

  cout << ans << '\n';
  return 0;
}
